Skip registry warning

jiayisuse opened the pull request:
Gloo device creator registry is throwing warning that confuses users.
Create C10_DEFINE_SHARED_REGISTRY_WITHOUT_WARNING API to skip such warning

Tested both `C10_DEFINE_SHARED_REGISTRY` and `C10_DEFINE_SHARED_REGISTRY_WITHOUT_WARNING`.
Make sure nothing breaks

d4l3k commented on c10/util/Registry.h:
```diff
@@ -91,7 +92,9 @@ class Registry {
             "Higher priority item already registered, skipping registration of " +
             KeyStrRepr(key);
-        fprintf(stderr, "%s\n", warn_msg.c_str());
+        if (warning_) {
```
Can you merge this if block with the parent else block?

d4l3k commented on c10/util/Registry.h:
```diff
@@ -153,6 +156,7 @@ class Registry {
   std::unordered_map<SrcType, Creator> registry_;
   std::unordered_map<SrcType, RegistryPriority> priority_;
   bool terminate_;
+  bool warning_;
```
nit make const

d4l3k commented:
Generally looks good, just a couple of nits

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 553de5d:


* **1/1** broken upstream at merge base 49fe7a7 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=49fe7a74014cd2aa803df14d547d0899c5582051&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=553de5dd728773bb473e622971222974ed371316).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=77816388)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

satgera commented:
Just to be sure CI failures aren't related, mind rebasing ?

smessmer reviewed (commented):
Code changes generally look good but the warning has a reason to be there, are you sure silencing it is the right way? Why do you need to overwrite already registered items? Maybe it's better to just avoid doing that?

jiayisuse committed:
Skip registry warning (#31126)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31126

Gloo device creator registry is throwing warning that confuses users - https://fb.workplace.com/groups/1405155842844877/permalink/3217491788277931/
Create C10_DEFINE_SHARED_REGISTRY_WITHOUT_WARNING API to skip such warning

Test Plan:
{F224342749}

Tested both `C10_DEFINE_SHARED_REGISTRY` and `C10_DEFINE_SHARED_REGISTRY_WITHOUT_WARNING`.
Make sure nothing breaks

Reviewed By: d4l3k

Differential Revision: D18904783

fbshipit-source-id: c88b95718d9de141aacf2d7ace3a9f977f5768cb

pietern commented:
@smessmer This can happen with runtime registration of higher priority entries. If the low priority (fallback) entry is registered first and the higher priority (preferred) entry later, the warning triggers.
